### DISCUSSION TITLE
Audio-Categorize Plugin

### DISCUSSION THREAD

**Discussion by glarn49** [jspsych/jsPsych#979] [Q&A]
This is my first time working with HTML/Javascript, so apologies in advance for a potentially silly question. I'm trying to modify the existing categorize-image plugin to play audio instead of display an image, but I'm running into some issues. When I run the task, everything works except the audio doesn't play. This is true across browsers and using different audio files. 

When running in Chrome, I get the warning "The AudioContext was not allowed to start. It must be resumed (or created) after a user gesture on the page." I've read elsewhere that this means you should have the user interact with the task before the audio trials, but my task does prompt user interaction with an instruction screen, so I'm not sure the context for that warning. My worry is that I'm missing something in the plugin.trial functions, since I've just stuck pieces together from existing plugins.   I've included that section below. Anything I might be missing there, or other workarounds for the missing audio? Thanks!

>   
    plugin.trial = function(display_element, trial) {

    // setup stimulus
    var context = jsPsych.pluginAPI.audioContext();
    if(context !== null){
      var source = context.createBufferSource();
      source.buffer = jsPsych.pluginAPI.getAudioBuffer(trial.stimulus);
      source.connect(context.destination);
    } else {
      var audio = jsPsych.pluginAPI.getAudioBuffer(trial.stimulus);
    }

    // set up end event if trial needs it

    if(trial.trial_ends_after_audio){
      if(context !== null){
        source.onended = function() {
          end_trial();
        }
      } else {
        audio.addEventListener('ended', end_trial);
      }
    }

    // if prompt is set, show prompt
    if (trial.prompt !== null) {
      display_element.innerHTML += trial.prompt;
    }

    var trial_data = {};

    // create response function
    var after_response = function(info) {

      // kill any remaining setTimeout handlers
      jsPsych.pluginAPI.clearAllTimeouts();

      // clear keyboard listener
      jsPsych.pluginAPI.cancelAllKeyboardResponses();

      var correct = false;
      if (trial.key_answer == info.key) {
        correct = true;
      }

      // save data
      trial_data = {
        "rt": info.rt,
        "correct": correct,
        "stimulus": trial.stimulus,
        "key_press": info.key
      };

      display_element.innerHTML = '';

      var timeout = info.rt == null;
      doFeedback(correct, timeout);
    }

    jsPsych.pluginAPI.getKeyboardResponse({
      callback_function: after_response,
      valid_responses: trial.choices,
      rt_method: 'performance',
      persist: false,
      allow_held_key: false
    });

    if (trial.trial_duration !== null) {
      jsPsych.pluginAPI.setTimeout(function() {
        after_response({
          key: null,
          rt: null
        });
      }, trial.trial_duration);
    }

    function doFeedback(correct, timeout) {

      if (timeout && !trial.show_feedback_on_timeout) {
        display_element.innerHTML += trial.timeout_message;
      } else {
        // show image during feedback if flag is set
        if (trial.show_stim_with_feedback) {
          display_element.innerHTML = '<img id="jspsych-categorize-image-stimulus" class="jspsych-categorize-image-stimulus" src="'+trial.stimulus+'"></img>';
        }

        // substitute answer in feedback string.
        var atext = "";
        if (correct) {
          atext = trial.correct_text.replace("%ANS%", trial.text_answer);
        } else {
          atext = trial.incorrect_text.replace("%ANS%", trial.text_answer);
        }

        // show the feedback
        display_element.innerHTML += atext;
      }
      // check if force correct button press is set
      if (trial.force_correct_button_press && correct === false && ((timeout && trial.show_feedback_on_timeout) || !timeout)) {

        var after_forced_response = function(info) {
          endTrial();
        }

        jsPsych.pluginAPI.getKeyboardResponse({
          callback_function: after_forced_response,
          valid_responses: [trial.key_answer],
          rt_method: 'performance',
          persist: false,
          allow_held_key: false
        });

      } else {
        jsPsych.pluginAPI.setTimeout(function() {
          endTrial();
        }, trial.feedback_duration);
      }

    }

    function endTrial() {
      display_element.innerHTML = '';
      jsPsych.finishTrial(trial_data);
    }

  };

**Comment by jodeleeuw** [comment]
Hi @glarn49 ,

It could be that the audio files aren't loaded. Did you preload the audio files anywhere? Either through the `preload_audio` option in `jsPsych.init()` or by using the `registerPreload` method in the plugin? 

Here's the relevant bit from the `audio-keyboard-response` plugin.

https://github.com/jspsych/jsPsych/blob/2cbd571a186dab97024e25d029950cfcc37c94f1/plugins/jspsych-audio-keyboard-response.js#L15
